File: BOJ/이분탐색/Q2110.py



N, C = map(int, input().split())
houses = list(int(input()) for _ in range(N))
houses.sort()

lt = 1
rt = houses[-1] - houses[0]
rst = rt

while lt <= rt:
    mid = (rt + lt) // 2


    cnt = 1
    start = houses[0]
    for i in range(1, N):
        distance = houses[i] - start
        if mid <= distance:
            cnt += 1
            start = houses[i]

    if cnt < C:
        rt = mid - 1
    else:
        rst = mid
        lt = mid + 1

print(rst)








'''
6 4
1
2
3
9
11
12
=> 2
1 12 6
1 5 3
1 2 1
2 2 2


5 3
1
2
8
4
9
=> 3
[1, 2, 4, 8, 9]



'''

File: Algorithm-java/src/BOJ/Gold/Q2110.java
package BOJ.Gold;

import java.io.*;
import java.util.Arrays;

public class Q2110 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        String[] input = br.readLine().trim().split(" ");
        int N = Integer.parseInt(input[0]);
        int C = Integer.parseInt(input[1]);

        int[] pos = new int[N];
        for (int i = 0; i < N; i++)
            pos[i] = Integer.parseInt(br.readLine().trim());

        Arrays.sort(pos);

        int lt = 1;
        int rt = pos[N - 1] - pos[0];
        int rst = 0;

        while (lt <= rt) {
            int mid = (lt + rt) / 2;

            int cnt = 1;
            int lastLocate = pos[0];
            for (int i = 0; i < N; i++) {
                int locate = pos[i];
                if (locate - lastLocate >= mid) {
                    cnt++;
                    lastLocate = locate;
                }
            }
            if (cnt < C) {
                rt = mid - 1;
            } else {
                rst = mid;
                lt = mid + 1;
            }
        }

        bw.write(String.valueOf(rst));
        bw.flush();
    }
}
